service/history/chasm_engine.go

```diff
@@ -241,14 +278,173 @@ func (e *ChasmEngine) ReadComponent(
+// history. Thus when the predicate function is evaluated, it is guaranteed that the entity VT >=
+// requestRef VT.
 func (e *ChasmEngine) PollComponent(
 	ctx context.Context,
-	entityRef chasm.ComponentRef,
-	predicateFn func(chasm.Context, chasm.Component) (any, bool, error),
-	operationFn func(chasm.MutableContext, chasm.Component, any) error,
-	opts ...chasm.TransitionOption,
-) (newEntityRef []byte, retError error) {
-	return nil, serviceerror.NewUnimplemented("PollComponent is not yet supported")
+	requestRef chasm.ComponentRef,
+	predicateFn func(chasm.Context, chasm.Component) (bool, error),
+) ([]byte, error) {
+	// 1. Acquire lock
+	// 2. Error if shard entity VT < requestRef VT ('stale state')
+	// 3. If predicate satisfied, release lock and return
+	// 4. Subscribe to notifications for this entity
+	// 5. Release lock
+	// 6. On notification repeat (1) and (2)
+
+	shardContext, executionLease, err := e.getExecutionLease(ctx, requestRef)
+	if err != nil {
+		// E.g. requestRef VT inconsistent with shard VT ('stale reference')
+		return nil, err
+	}
+
+	// hold lock until return or subscription established
+	released := false
+	defer func() {
+		if !released {
+			// read-only operation; don't pass error
+			executionLease.GetReleaseFn()(nil)
+		}
+	}()
+
+	// At this point it's possible that shard VT < requestRef VT (getExecutionLease does not
+	// guarantee that returned shard state is non-stale w.r.t. requestRef).
+	currentRef, err := e.checkPredicate(ctx, requestRef, executionLease, predicateFn)
+	if err != nil {
+		return nil, err
+	}
+	// shard VT >= requestRef VT (enforced by checkPredicate)
+
+	if currentRef != nil {
+		// wait condition was satisfied
+		return currentRef, nil
+	}
+
+	// Wait condition not satisfied; long-poll
+
+	channel, subscriberID, err := e.notifier.Subscribe(requestRef.EntityKey)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = e.notifier.Unsubscribe(requestRef.EntityKey, subscriberID)
+	}()
+
+	// Release the lock, now that we are subscribed
+	executionLease.GetReleaseFn()(nil)
+	released = true
+
+	namespaceRegistry, err := shardContext.GetNamespaceRegistry().GetNamespaceByID(
+		namespace.ID(requestRef.EntityKey.NamespaceID),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Long-poll timeout semantics are the same as PollWorkflowExecutionUpdate: if the long-poll
+	// times out due to the server-imposed soft timeout, then return a non-error empty response
+	// indicating to a caller that they should continue long-polling.
+
+	// TODO(dan): GetOrPollMutableState uses
+	// longPollInterval := shardContext.GetConfig().LongPollExpirationInterval(namespaceRegistry.Name().String())
+	// longPollCtx, cancel := contextutil.WithDeadlineBuffer(ctx, longPollInterval, longPollSoftTimeout)
+	softTimeout := shardContext.GetConfig().LongPollExpirationInterval(namespaceRegistry.Name().String())
+	stCtx, stCancel := context.WithTimeout(ctx, softTimeout)
+	defer stCancel()
```

Comment:
I'm still thinking about the timeouts and time budgeting here.

We are potentially deep in the stack here. I wonder whether the `softTimeout` needs to be passed into this function.

Reply:
What you put here LGTM.